compiler/hash-typecheck/src/new/passes/symbol_resolution.rs

```diff
@@ -120,8 +276,644 @@ impl<'tc> SymbolResolutionPass<'tc> {
     }
 }
 
-/// @@Temporary: for now this visitor just walks the AST and enters scopes. The
-/// next step is to resolve symbols in these scopes!.
+/// An argument group in the AST.
+///
+/// This is either a group of explicit `(a, b, c)` arguments, or a group of
+/// implicit `<a, b, c>` arguments. The former corresponds to the
+/// [`ast::ConstructorCallArg`], while the latter corresponds to the
+/// [`ast::TyArg`].
+#[derive(Copy, Clone, Debug)]
+#[allow(dead_code)]
+enum AstArgGroup<'a> {
+    /// A group of explicit `(a, b, c)` arguments.
+    ExplicitArgs(&'a ast::AstNodes<ast::ConstructorCallArg>),
+    /// A group of implicit `<a, b, c>` arguments.
+    ImplicitArgs(&'a ast::AstNodes<ast::TyArg>),
+}
```

Comment:
I don't quite understand the naming conventions, why is `(a, b, c)` "explicit", but `<a, b, c>` is "implicit"?

Reply:
As in `<a, b, c>` is optional so you don't have to necessarily provide it and it will be implicitly inferred if you don't

Reply:
But default arguments would produce a similar effect when you don't specify them... they would also be "implied"

Reply:
Yeah, but that is only for specific arguments in the explicit args, where as a `<..>` group can be fully inferred

Reply:
Would it not be better to call it something like `ValueArgs` and `TyArgs`?

Reply:
No, because that's not what they are; types and values are valid in both arguments.

Reply:
But `ast::ConstructorCallArg` must have a value present I believe?

Reply:
Yes, as far as the AST goes it expects values for the explicit parameters and types for the implicit parameters, however as far as the typechecker is concerned they both accept the same thing, since `Type` is a valid type. I am planning on changing the parsing a bit to also allow `(..)` in types and `<..>` in expressions to make it more uniform. Most of the time, `<..>` will be used for types and `(..)` will be used for values, but this is simply because you want types to usually be inferred, and values not. Also by default untyped parameters `<T>` will be assumed to be `<T: Type>` for implicit functions, while this is not the case for explicit functions.